src/mrinufft/operators/interfaces/gpunufft.py

```diff
@@ -542,3 +545,103 @@ def _safe_squeeze(self, arr):
             except ValueError:
                 pass
         return arr
+
+    def data_consistency(self, image_data, obs_data):
+        """Compute the data consistency estimation directly on gpu.
+
+        This mixes the op and adj_op method to perform F_adj(F(x-y))
+        on a per coil basis. By doing the computation coil wise,
+        it uses less memory than the naive call to adj_op(op(x)-y)
+
+        Parameters
+        ----------
+        image: array
+            Image on which the gradient operation will be evaluated.
+            N_coil x Image shape is not using sense.
+        obs_data: array
+            Observed data.
+        """
+        xp = get_array_module(image_data)
+        if xp.__name__ == "torch":
+            image_data_device = image_data.device
+            if image_data.is_cpu:
+                image_data = image_data.numpy()
+            else:
+                image_data = cp.from_dlpack(image_data)
+        xp = get_array_module(obs_data)
+        if xp.__name__ == "torch":
+            if obs_data.is_cpu:
+                obs_data = obs_data.numpy()
+            else:
+                obs_data = cp.from_dlpack(obs_data)
+
+        obs_data = auto_cast(obs_data, self.cpx_dtype)
+        image_data = auto_cast(image_data, self.cpx_dtype)
+
+        B, C = self.n_batchs, self.n_coils
+        K, XYZ = self.n_samples, self.shape
+
+        check_size(obs_data, (B, C, K))
+        if self.uses_sense:
+            check_size(image_data, (B, *XYZ))
+        else:
+            check_size(image_data, (B, C, *XYZ))
+
+        # dispatch
+        if is_host_array(image_data) and is_host_array(obs_data):
+            grad_func = self._dc_host
+        elif is_cuda_array(image_data) and is_cuda_array(obs_data):
+            grad_func = self._dc_direct
+        else:
+            raise ValueError("image and obs_data should be on the same device")
+
+        if not self.impl.use_gpu_direct:
```

Comment:
To me this warning doesnt make sense here. This is should be reported only if the input is a GPU array